### PR TITLE
Fix LateBoundMDC instantiation

### LINKED ISSUE #26258: OpenTelemetry MDC integration not quite working
### Describe the bug

While Quarkus appears to register the trace context fine in the MDC using `VertxMDC`, it appears that it cannot be accessed using the `org.jboss.logging.MDC` class. AFAICT the issue appears to be that the static initialization of `org.jboss.logmanager.MDC#mdcProvider` may not be able to find the `io.quarkus.bootstrap.logging.LateBoundMDCProvider` as registered in a corresponding `META-INF/services` file. This appears to be due to the fact that the mentioned field gets initialized too early and not explicitly using the TCCL. At least when I start my application in dev-mode and set a breakpoint in `org.jboss.logmanager.MDC#getMDCProvider()` the class initialization is triggered by the initialization of `io.quarkus.deployment.dev.DevModeMain#log`.

In my "ultimate" use case I simply want to use the `quarkus-logging-json` extension to log the log records (including MDC) in JSON and that should include the MDC. But when the logger attempts to get the MDC using `org.jboss.logmanager.ExtLogRecord#getMdcCopy()` (which calls `org.jboss.logmanager.MDC.copyObject()`) the returned result is always empty, since it ends up using the default `org.jboss.logmanager.ThreadLocalMDC` rather than the `io.quarkus.bootstrap.logging.LateBoundMDCProvider` implementation.

### Expected behavior

The MDC data supplied by the OpenTelemetry integration should be available via the `org.jboss.logging.MDC` API.

### Actual behavior

When using the `org.jboss.logging.MDC` API and also the `quarkus-logging-json` extension the MDC data supplied by the Quarkus OpenTelemetry integration is not available.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### LINKED ISSUE #26486: Values set with org.slf4j.MDC no longer works in 2.10.0.Final
### Describe the bug

I recently upgrade my shared component from Quarkus 2.9.2 to 2.10.0 where I previously had a logging filter that used `org.slf4j.MDC` to set context but my tests for it started failing.

Once I switched to use `org.jboss.logging.MDC` then it started working.

### Expected behavior

Given that the logging is supposed to merge the SLF4J calls I'd have expected calls to `org.slf4j.MDC` to work just fine.

### Actual behavior

The context was not propagated properly until I switched to `org.jboss.logging.MDC`.

### How to Reproduce?

Here's the sample code:

```
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import javax.annotation.PostConstruct;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerRequestFilter;
import javax.ws.rs.container.ContainerResponseContext;
import javax.ws.rs.container.ContainerResponseFilter;
import javax.ws.rs.ext.Provider;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.MDC;

@Provider
public class LoggingFilter implements ContainerRequestFilter, ContainerResponseFilter {
    // Made this default scope so that we can see it from the tests.
    static final List<String> DEFAULT_HEADERS_TO_LOG =
            Arrays.asList(
                    "TRACKING-ID", "SESSION-ID");

    @ConfigProperty(name = "logging.additional.headers")
    Optional<List<String>> additionalHeaders;

    private final List<String> allHeaders;

    /**
     * Default constructor used when this is called by the servlet container when its in {@code
     * web.xml}.
     */
    public LoggingFilter() {
        allHeaders = new ArrayList<>();
        allHeaders.addAll(DEFAULT_HEADERS_TO_LOG);
    }

    /** PostConstruct that checks to see if additional headers were populated. */
    @PostConstruct
    public void init() {
        additionalHeaders.ifPresent(allHeaders::addAll);
    }

    String convertListToString(List<String> list) {
        if (list == null || list.isEmpty()) {
            return "UNKNOWN";
        }

        return list.stream().collect(Collectors.joining(","));
    }

    @Override
    public void filter(ContainerRequestContext requestContext) throws IOException {
        // We're assuming these headers have a singular value
        allHeaders.forEach(
                name -> MDC.put(name, convertListToString(requestContext.getHeaders().get(name))));
    }

    @Override
    public void filter(
            ContainerRequestContext requestContext, ContainerResponseContext responseContext)
            throws IOException {
        MDC.clear();
    }
}
```

and the associated unit test:

```
import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerResponseContext;
import javax.ws.rs.core.MultivaluedHashMap;
import javax.ws.rs.core.MultivaluedMap;
import org.apache.commons.lang3.reflect.FieldUtils;
import org.slf4j.MDC;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

class LoggingFilterTest {
    private static final MultivaluedMap<String, String> MULTIVALUED_HASH_MAP =
            new MultivaluedHashMap();

    @BeforeEach
    void setup() {
        MULTIVALUED_HASH_MAP.add("One", "One");
    }

    /** Test with no additional details but an HTTP request. */
    @Test
    void doFilterHttpNoAdditional() throws IOException, IllegalAccessException {
        var requestContext = Mockito.mock(ContainerRequestContext.class);
        var responseContext = Mockito.mock(ContainerResponseContext.class);

        LoggingFilter.DEFAULT_HEADERS_TO_LOG.forEach(
                name -> Mockito.when(requestContext.getHeaders()).thenReturn(MULTIVALUED_HASH_MAP));

        var filter = new LoggingFilter();
        FieldUtils.writeField(filter, "additionalHeaders", Optional.empty(), true);
        filter.init();

        filter.filter(requestContext);

        assertThat(MDC.getMap())
                .isEqualTo(
                        Map.of(
                                "SESSION-ID",
                                "UNKNOWN",
                                "TRACKING-ID",
                                "UNKNOWN"));

        filter.filter(requestContext, responseContext);

        assertThat(MDC.getMap()).isEqualTo(Collections.emptyMap());

        Mockito.verifyNoInteractions(responseContext);
    }

    /** Test filter config */
    @Test
    void doFilterHttpFilterConfig() throws IOException, IllegalAccessException {
        var requestContext = Mockito.mock(ContainerRequestContext.class);
        var responseContext = Mockito.mock(ContainerResponseContext.class);

        var values = Optional.of(Arrays.asList("test", "one", "two"));

        var verifyHeaders = new ArrayList<String>();
        verifyHeaders.addAll(LoggingFilter.DEFAULT_HEADERS_TO_LOG);
        verifyHeaders.addAll(values.get());
        verifyHeaders.forEach(
                name -> Mockito.when(requestContext.getHeaders()).thenReturn(MULTIVALUED_HASH_MAP));

        var filter = new LoggingFilter();
        FieldUtils.writeField(filter, "additionalHeaders", values, true);
        filter.init();

        filter.filter(requestContext);

        assertThat(MDC.getMap())
                .isEqualTo(
                        Map.of(
                                "SESSION-ID",
                                "UNKNOWN",
                                "TRACKING-ID",
                                "UNKNOWN",
                                "one",
                                "UNKNOWN",
                                "test",
                                "UNKNOWN",
                                "two",
                                "UNKNOWN"));

        filter.filter(requestContext, responseContext);

        assertThat(MDC.getMap()).isEqualTo(Collections.emptyMap());

        Mockito.verifyNoInteractions(responseContext);
    }

    /** Test enumeration conversion. */
    @Test
    void convertListToString() {
        var filter = new LoggingFilter();
        var list = Arrays.asList("One", "Two", "Three");

        var response = filter.convertListToString(list);
        assertThat(response).isEqualTo("One,Two,Three");
    }

    /** Test enumeration conversion. */
    @Test
    void convertEmptyEnumerationToString() {
        var filter = new LoggingFilter();

        var response = filter.convertListToString(Collections.emptyList());
        assertThat(response).isEqualTo("UNKNOWN");
    }

    /** Test enumeration conversion. */
    @Test
    void convertNullEnumerationToString() {
        var filter = new LoggingFilter();

        var response = filter.convertListToString(null);
        assertThat(response).isEqualTo("UNKNOWN");
    }
}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk version "11.0.15" 2022-04-19 LTS OpenJDK Runtime Environment Corretto-11.0.15.9.1 (build 11.0.15+9-LTS) OpenJDK 64-Bit Server VM Corretto-11.0.15.9.1 (build 11.0.15+9-LTS, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by luneo7**
Fixes https://github.com/quarkusio/quarkus/issues/26258

Move the service provider configuration to the bootstrap project, and make the default behavior of LateBoundMDCProvider to be a ThreadLocalMDC if no delegate is set, this way if no delegate is used the MDC behaves as the traditional ThreadLocalMDC, and if one is set the behavior will be of the delegate.

### CODE REVIEW

**Comment by geoand** [comment]
Thanks for this, but I don't think we always want to have the service active

**Comment by geoand** [comment]
Thanks a lot!

This indeed is close to what I had in mind.
But for the sake of efficiency, we don't need an extra jar like you have added - we can simply generate the parent first resources inside `quarkus-run.jar`.

For this to work in dev-mode, there will likely need to be changes in `StartupActionImpl`. 
Would you like to continue this investigation?

**Comment by luneo7** [comment]
And moving this to a build item, will also have to fix test mode... since there is no provider config physical resource, the classloader there also loads MDC before the generated resource is known...

**Comment by geoand** [comment]
We would need to alter the dev-mode jar to make this work in dev-mode... That cannot be done with build items unfortunately.
@aloubyansky would it make sense to add some kind of extension configuration that would allows for the altering of the dev-mode jar?

**Comment by aloubyansky** [comment]
In which way @geoand?

**Comment by geoand** [comment]
We basically want to add a services file to the dev-mode jar (when an extension exists)

**Comment by aloubyansky** [comment]
Does test mode have the same issue?

**Comment by geoand** [comment]
Good point... It does :(

**Comment by geoand** [comment]
Fast-jar and dev-mode are possible. Test mode would not work because you have no control of the created jar that launches the tests

**Comment by luneo7** [comment]
Another option... is just creating a simple maven project inside the vertx project (dunno `latebound-mdc-provider`) and then when importing that lib in the vertx runtime we just mark it as a parent first ? This seems to be the simpler solution to me

**Comment by geoand** [comment]
Pretty sure that would not work, but you are certainly welcome to try

**Comment by aloubyansky** [comment]
I think it could work if the interfaces can be moved to system classloader too.

**Comment by luneo7** [comment]
I will give a quick try =]

**Comment by luneo7** [comment]
Creating a module inside vertx does the trick, dev mode, test mode, fast-jar, and all modes/packages works as expected =] @geoand do you think this is something that we want?

**Comment by aloubyansky** [comment]
@luneo7 could you show us the changes you tried?

**Comment by luneo7** [comment]
What I have tried before [d9b0c40](https://github.com/quarkusio/quarkus/pull/26298/commits/d9b0c406646be58bc9c82bbce85397b007440845):
- Custom build item (didn't work for dev mode and test mode)
- Tried to change the maven extension... but as @geoand said we have no control of the test jar... so it didn't work
What worked, was creating a lib inside vertx package, that is imported by vertx core runtime and in vertx core runtime we set the parent first config there in the pom file, so it works for dev mode, test mode, fast jar and everything else

**Comment by aloubyansky** [comment]
Oh, sorry. Looks pretty clean to me.

**Comment by luneo7** [comment]
Improved the unit testing, added an integration testing and a dev mode test =]
Going to convert this to a regular pull request

**Comment by luneo7** [comment]
@aloubyansky @geoand the PR is ready for review. I think it would be a good thing to backport it to 2.10

File: extensions/vertx/deployment/src/test/java/io/quarkus/vertx/mdc/VertxMDCDevModeTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                    (jar) -> jar.addClasses(JavaArchive.class, VerticleDeployer.class, InMemoryLogHandler.class,
+                            InMemoryLogHandlerProducer.class)
+                            .add(new StringAsset("quarkus.log.file.enable=true\n" +
+                                    "quarkus.log.console.format=%d{HH:mm:ss} %-5p requestId=%X{requestId} [%c{2.}] (%t) %s%e%n\n"),
+                                    "application.properties"))
+            .setLogFileName("quarkus-mdc.log");
+
+    @Test
+    void mdcDevMode() {
+        Path logDirectory = Paths.get(".", "target");
```
Comment (aloubyansky) [inline extensions/vertx/deployment/src/test/java/io/quarkus/vertx/mdc/VertxMDCDevModeTest.java:37]: It's equivalent to simply `Paths.get("target")` which in its turn calls `Path.of("target")`. You don't have to change it though.

**Review by aloubyansky — APPROVED** [review]
Thanks @luneo7!

**Comment by luneo7** [comment]
That TCK test, seems flaky... locally it went through... also saw that in some other PRs the same test is failing

image

**Comment by geoand** [comment]
The TCK failure is addressed by #26492

**Review by geoand — APPROVED** [review]
Thanks!

I am not sold on backporting this however - it seems a little too risky for backporting.
I'll let @gsmet decide

**Comment by gsmet** [comment]
> I am not sold on backporting this however - it seems a little too risky for backporting.

Yeah, same here. I'm not comfortable backporting it.

**Comment by luneo7** [comment]
Whatever you guys decide, there are complaints about it in 2.10 though, so people would have to wait for 2.11 to have it fixed

**Pull request merged**